src/stream.jl

```diff
@@ -310,16 +326,18 @@ end
 # --------------
 
 function Base.read(stream::TranscodingStream, ::Type{UInt8})
-    ready_to_read!(stream)
     if eof(stream)
-        throw(EOFError())
+        ready_to_read!(stream)
```

Comment:
I think `ready_to_read!(stream)` needs to be called before `eof(stream)` to error in case the stream is in `:write` mode.

For example with this PR, the following happens.

```julia
julia> using TranscodingStreams, CodecZlib

julia> stream = IOBuffer(zeros(UInt8,10); write=true, read=true)
IOBuffer(data=UInt8[...], readable=true, writable=true, seekable=true, append=false, size=10, maxsize=Inf, ptr=1, mark=-1)

julia> eof(stream)
false

julia> stream = TranscodingStream(GzipCompressor(), stream)
TranscodingStream{GzipCompressor, IOBuffer}(<mode=idle>)

julia> write(stream, "hi")
2

julia> read(stream, UInt8)
0x68
```
The `read` after `write` used to throw 
```julia
ERROR: ArgumentError: cannot change the mode from write to read
Stacktrace:
 [1] changemode!(stream::TranscodingStream{GzipCompressor, IOBuffer}, newmode::Symbol)
   @ TranscodingStreams ~/.julia/packages/TranscodingStreams/LZb0O/src/stream.jl:777
 [2] ready_to_read!
   @ ~/.julia/packages/TranscodingStreams/LZb0O/src/stream.jl:442 [inlined]
 [3] read(stream::TranscodingStream{GzipCompressor, IOBuffer}, ::Type{UInt8})
   @ TranscodingStreams ~/.julia/packages/TranscodingStreams/LZb0O/src/stream.jl:313
 [4] top-level scope
   @ REPL[14]:1
```

Reply:
That seems like an unrelated bug, since calling `eof` should not return false if there are no bytes to read

Reply:
But anyways, we can manually inline the `eof` check here, it just seems very sketchy to require doing that

Reply:
Yes, that makes sense to me. There is an unrelated bug with `eof` going on here. From the docs of `eof`

> Therefore it is always safe to read one byte after seeing
  eof return false.

So `eof` must not return `false` if it is an error to read a byte.

Reply:
The converse requirement is also common too, but less strict: if `eof` returns `true` generally it shouldn't later return `false` (there are some exceptions though, as seeking or writing to a file or iobuffer or calling `reseteof` for some streams will reset that indicator)

Reply:
Hmm, I guess this would happen if the stream goes from `:write` mode to `:stop` mode, which I guess could happen if the `stop_on_end` flag is used before writing. Maybe that should be an error.

Reply:
Fixes for `eof` behavior are complex enough to be in a separate PR. For this PR it would be good to keep the `ready_to_read!(stream)` at the start of this function to keep the current behavior.

Reply:
We actually have to reimplement both of those code in this function then. Not hard, just really awkward for that to be necessary

Reply:
Alright, updated to just inline the fixes for those functions rather than use them. Performance should be the same